### PR TITLE
SVM: Clarify usage of `lamports_per_signature`

### PR DESCRIPTION
**Pull request by buffalojoec**
#### Problem
SVM uses two values for `lamports_per_signature` - one from the `FeeStructure` and one from the blockhash queue (fee governor). The former is used to actually assess transaction fees.

The relationship between these two `lamports_per_signature` values is not clear from the SVM API's documentation or symbol names.

#### Summary of Changes
First remove the entire `FeeStructure` type and just use its `lamports_per_signature` value directly, since SVM has no need for the rest of the type.

Then, rename the `lamports_per_signature` that comes from the blockhash queue (fee governor) to `blockhash_lamports_per_signature` in both SVM and the program-runtime.

### CODE REVIEW

File: program-runtime/src/invoke_context.rs [hunk 1]
```diff
@@ -147,27 +147,27 @@ impl BpfAllocator {
...
     epoch_total_stake: Option<u64>,
     epoch_vote_accounts: Option<&'a VoteAccountsHashMap>,
     pub feature_set: Arc<FeatureSet>,
-    pub lamports_per_signature: u64,
     sysvar_cache: &'a SysvarCache,
 }
 impl<'a> EnvironmentConfig<'a> {
     pub fn new(
         blockhash: Hash,
+        blockhash_lamports_per_signature: u64,
```
Comment (apfitzge) [inline program-runtime/src/invoke_context.rs:162]: why'd we move it?

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -3669,16 +3669,17 @@ impl Bank {
...
             RentCollectorWithMetrics::new(self.rent_collector.clone());
         let processing_environment = TransactionProcessingEnvironment {
             blockhash,
+            blockhash_lamports_per_signature,
             epoch_total_stake: self.epoch_total_stake(self.epoch()),
             epoch_vote_accounts: self.epoch_vote_accounts(self.epoch()),
             feature_set: Arc::clone(&self.feature_set),
-            fee_structure: Some(&self.fee_structure),
-            lamports_per_signature,
+            fee_per_signature: self.fee_structure.lamports_per_signature,
```
Comment (apfitzge) [inline runtime/src/bank.rs]: should we call it `fee_lamports_per_signature`?

File: svm/src/transaction_processor.rs [hunk 3]
```diff
@@ -124,16 +124,21 @@ pub struct TransactionProcessingConfig<'a> {
 pub struct TransactionProcessingEnvironment<'a> {
     /// The blockhash to use for the transaction batch.
     pub blockhash: Hash,
+    /// Lamports per signature that corresponds to this blockhash.
+    ///
+    /// Note: This value is primarily used for nonce accounts. If set to zero,
+    /// it will disable transaction fees. However, any non-zero value will not
+    /// change transaction fees. For this reason, it is recommended to use the
+    /// `fee_per_signature` field to adjust transaction fees.
+    pub blockhash_lamports_per_signature: u64,
```
Comment (apfitzge) [inline svm/src/transaction_processor.rs:133]: is this used for anything other than setting fees to zero?
I'm trying to remove that functionality now in #3211.

Is this exposed to user tx somehow? i.e. can user programs get this from some syscall?

File: program-runtime/src/invoke_context.rs (hunk 1 above)
Comment (buffalojoec) [reply program-runtime/src/invoke_context.rs:162]: Alpha-ordering 😅

File: runtime/src/bank.rs (hunk 2 above)
Comment (buffalojoec) [reply runtime/src/bank.rs]: I thought about it, but I thought that sounded weird tbh.

File: svm/src/transaction_processor.rs (hunk 3 above)
Comment (buffalojoec) [reply svm/src/transaction_processor.rs:133]: It's used in the `InvokeContext` as a queryable value for System program nonce operations.
https://github.com/anza-xyz/agave/blob/c3693a6d33e34025da0070d572036c747a91f0f9/programs/system/src/system_instruction.rs#L59

I saw your PR. I was going to follow this up with another PR or two to start evicting the weird boolean check, but seems like you beat me to it.

I don't really know how it's used for nonces yet, though. I was inferring that it was locking nonce fee payments to the LPS at the time of the nonce account creation (or advancing), but it seems like SVM will still use `FeeStructure` for the fee calculation anyway.
https://github.com/anza-xyz/agave/blob/c3693a6d33e34025da0070d572036c747a91f0f9/svm/src/transaction_processor.rs#L470-L482

File: runtime/src/bank.rs (hunk 2 above)
Comment (buffalojoec) [reply runtime/src/bank.rs]: I actually decided this was a better name.

**Comment by buffalojoec** [comment]
@apfitzge this is ready for another look! With the revert of the removal of the `lamports_per_signature == 0` flag, I think this change is even more relevant.

File: svm/src/transaction_processor.rs [hunk 4]
```diff
@@ -121,24 +121,42 @@ pub struct TransactionProcessingConfig<'a> {
...
+    pub fee_lamports_per_signature: u64,
     /// Rent collector to use for the transaction batch.
     pub rent_collector: Option<&'a dyn SVMRentCollector>,
 }
 
+impl Default for TransactionProcessingEnvironment<'_> {
+    fn default() -> Self {
+        Self {
+            blockhash: Hash::default(),
+            blockhash_lamports_per_signature: 0,
```
Comment (apfitzge) [inline svm/src/transaction_processor.rs:150]: default using 0 means fees are zero.

Is this only used in tests/benchmarks? if so let's make the impl feature-dependent on dcou

File: svm/src/transaction_processor.rs [hunk 5]
```diff
@@ -121,24 +121,42 @@ pub struct TransactionProcessingConfig<'a> {
...
     /// The vote accounts for the current epoch.
     pub epoch_vote_accounts: Option<&'a VoteAccountsHashMap>,
     /// Runtime feature set to use for the transaction batch.
     pub feature_set: Arc<FeatureSet>,
-    /// Fee structure to use for assessing transaction fees.
-    pub fee_structure: Option<&'a FeeStructure>,
-    /// Lamports per signature to charge per transaction.
-    pub lamports_per_signature: u64,
+    /// Transaction fee to charge per signature, in lamports.
+    pub fee_lamports_per_signature: u64,
```
Comment (apfitzge) [inline svm/src/transaction_processor.rs:141]: wondering if we even need FeeStructure at all in `Bank` after this change. out of scope for this PR

Comment (buffalojoec) [reply svm/src/transaction_processor.rs:141]: Looks like maybe not. Can check it out for next maybe. I think the value it produces for LPS on `FeeStructure::default()` is the only crucial bit at this point.

File: svm/src/transaction_processor.rs (hunk 4 above)
Comment (buffalojoec) [reply svm/src/transaction_processor.rs:150]: This struct was already deriving `Default`, I just made it manual. Zero was already the value. I can move `Default` to dcou, but wondering if it's a strong enough case for removing an already implemented trait.

Comment (apfitzge) [reply svm/src/transaction_processor.rs:150]: Ah I just missed the original impl being derived.

Would still be better imo for this to be dcou only, but separate from this work.

**Review by apfitzge — APPROVED** [review]

**Pull request merged**